primitives/state-machine/src/overlayed_changes.rs

```diff
@@ -342,6 +488,67 @@ impl OverlayedChanges {
 		}
 	}
 
+	/// Generate the storage root using `backend` and all changes from `prospective` and `committed`.
+	///
+	/// Returns the storage root and caches storage transaction in the given `cache`.
+	pub fn storage_root<H: Hasher, N: BlockNumber, B: Backend<H>>(
+		&self,
+		backend: &B,
+		cache: &mut StorageTransactionCache<B::Transaction, H, N>,
+	) -> H::Out
+		where H::Out: Ord + Encode,
+	{
+		let child_storage_keys = self.prospective.children.keys()
+			.chain(self.committed.children.keys());
+		let child_delta_iter = child_storage_keys.map(|storage_key|
+			(storage_key.clone(), self.committed.children.get(storage_key)
+				.into_iter()
+				.flat_map(|map| map.iter().map(|(k, v)| (k.clone(), v.value.clone())))
+				.chain(self.prospective.children.get(storage_key)
+					.into_iter()
+					.flat_map(|map| map.iter().map(|(k, v)| (k.clone(), v.value.clone()))))));
+
+		// compute and memoize
+		let delta = self.committed.top.iter().map(|(k, v)| (k.clone(), v.value.clone()))
```

Comment:
Do we really need to `clone` every key and value? Why can't `full_storage_root` work with borrowed values?

Reply:
I don't know, I don't have written this function. I think that I looked into this and there was no easy to do this.